# HG changeset patch
# Parent  80fb374d7109e03c44f049a2d7a845abf00663e1

diff --git a/lifx/wire_proto.c b/lifx/wire_proto.c
--- a/lifx/wire_proto.c
+++ b/lifx/wire_proto.c
@@ -382,7 +382,7 @@
     pkt->brightness = htole16(pkt->brightness);
     pkt->kelvin = htole16(pkt->kelvin);
     pkt->period = htole32(pkt->period);
-    pkt->cycles = lifx_wire_htolefloat(pkt->cycles);
+    pkt->cycles = lgtd_lifx_wire_htolefloat(pkt->cycles);
     pkt->skew_ratio = htole16(pkt->skew_ratio);
 }
 
diff --git a/lifx/wire_proto.h b/lifx/wire_proto.h
--- a/lifx/wire_proto.h
+++ b/lifx/wire_proto.h
@@ -26,14 +26,14 @@
 typedef float    floatle_t;
 
 static inline floatle_t
-lifx_wire_htolefloat(float f)
+lgtd_lifx_wire_htolefloat(float f)
 {
     *(uint32_t *)&f = htole32(*(uint32_t *)&f);
     return f;
 }
 
 static inline floatle_t
-lifx_wire_lefloattoh(float f)
+lgtd_lifx_wire_lefloattoh(float f)
 {
     *(uint32_t *)&f = le32toh(*(uint32_t *)&f);
     return f;
diff --git a/tests/core/proto/test_proto_set_waveform.c b/tests/core/proto/test_proto_set_waveform.c
--- a/tests/core/proto/test_proto_set_waveform.c
+++ b/tests/core/proto/test_proto_set_waveform.c
@@ -37,7 +37,7 @@
     int brightness = le16toh(waveform->brightness);
     int kelvin = le16toh(waveform->kelvin);
     int period = le32toh(waveform->period);
-    float cycles = lifx_wire_lefloattoh(waveform->cycles);
+    float cycles = lgtd_lifx_wire_lefloattoh(waveform->cycles);
     int skew_ratio = le16toh(waveform->skew_ratio);
 
     if (waveform_type != LGTD_LIFX_WAVEFORM_SAW) {
diff --git a/tests/core/proto/test_proto_set_waveform_on_routing_error.c b/tests/core/proto/test_proto_set_waveform_on_routing_error.c
--- a/tests/core/proto/test_proto_set_waveform_on_routing_error.c
+++ b/tests/core/proto/test_proto_set_waveform_on_routing_error.c
@@ -37,7 +37,7 @@
     int brightness = le16toh(waveform->brightness);
     int kelvin = le16toh(waveform->kelvin);
     int period = le32toh(waveform->period);
-    float cycles = lifx_wire_lefloattoh(waveform->cycles);
+    float cycles = lgtd_lifx_wire_lefloattoh(waveform->cycles);
     int skew_ratio = le16toh(waveform->skew_ratio);
 
     if (waveform_type != LGTD_LIFX_WAVEFORM_SAW) {
