# HG changeset patch
# Parent  298fe56d98508afd0b2d3098264f4661d5908fe5

diff --git a/core/lightsd.c b/core/lightsd.c
--- a/core/lightsd.c
+++ b/core/lightsd.c
@@ -138,6 +138,7 @@
         "  [-v,--verbosity debug|info|warning|error]\n",
         progname
     );
+    lgtd_cleanup();
     exit(0);
 }
 
@@ -159,6 +160,10 @@
     };
     const char short_opts[] = "l:c:fthv:V";
 
+    if (argc == 1) {
+        lgtd_usage(argv[0]);
+    }
+
     for (int rv = getopt_long(argc, argv, short_opts, long_opts, NULL);
          rv != -1;
          rv = getopt_long(argc, argv, short_opts, long_opts, NULL)) {
diff --git a/core/version.h.in b/core/version.h.in
--- a/core/version.h.in
+++ b/core/version.h.in
@@ -29,4 +29,7 @@
 
 #pragma once
 
-const char LGTD_VERSION[] = "@LGTD_VERSION@";
+const char LGTD_VERSION[] = (
+    "@LIGHTSD_VERSION@"
+    "\n\nCopyright (c) 2014, 2015, Louis Opter <kalessin@kalessin.fr>"
+);
