# HG changeset patch
# Parent  0a70598874618551b57f92c9f8fb83769a04487f
UI/Cosmetic fixes

diff --git a/core/lightsd.c b/core/lightsd.c
--- a/core/lightsd.c
+++ b/core/lightsd.c
@@ -139,6 +139,7 @@
         "  [-v,--verbosity debug|info|warning|error]\n",
         progname
     );
+    lgtd_cleanup();
     exit(0);
 }
 
@@ -162,6 +163,10 @@
     };
     const char short_opts[] = "l:c:fthv:V";
 
+    if (argc == 1) {
+        lgtd_usage(argv[0]);
+    }
+
     for (int rv = getopt_long(argc, argv, short_opts, long_opts, NULL);
          rv != -1;
          rv = getopt_long(argc, argv, short_opts, long_opts, NULL)) {
@@ -205,7 +210,7 @@
             }
             break;
         case 'V':
-            printf("%s v%s\n", argv[0], LGTD_VERSION);
+            printf("lightsd v%s\n", LGTD_VERSION);
             return 0;
         default:
             lgtd_usage(argv[0]);
diff --git a/core/version.h.in b/core/version.h.in
--- a/core/version.h.in
+++ b/core/version.h.in
@@ -29,4 +29,7 @@
 
 #pragma once
 
-const char LGTD_VERSION[] = "@LGTD_VERSION@";
+const char LGTD_VERSION[] = (
+    "@LIGHTSD_VERSION@\n\n"
+    "Copyright (c) 2014, 2015, Louis Opter <kalessin@kalessin.fr>"
+);
