# HG changeset patch
# Parent  f6d85d8c6ff6f643fb22fa304f67c35d942266e1

diff --git a/tests/lifx/gateway/test_gateway_allocate_tag_id.c b/tests/lifx/gateway/test_gateway_allocate_tag_id.c
--- a/tests/lifx/gateway/test_gateway_allocate_tag_id.c
+++ b/tests/lifx/gateway/test_gateway_allocate_tag_id.c
@@ -4,6 +4,7 @@
 
 #define MOCKED_LIFX_TAGGING_INCREF
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 static bool tagging_incref_called = false;
 
diff --git a/tests/lifx/gateway/test_gateway_allocate_tag_id_from_lifx_network.c b/tests/lifx/gateway/test_gateway_allocate_tag_id_from_lifx_network.c
--- a/tests/lifx/gateway/test_gateway_allocate_tag_id_from_lifx_network.c
+++ b/tests/lifx/gateway/test_gateway_allocate_tag_id_from_lifx_network.c
@@ -4,6 +4,7 @@
 
 #define MOCKED_LIFX_TAGGING_INCREF
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 static bool tagging_incref_called = false;
 
diff --git a/tests/lifx/gateway/test_gateway_allocate_tag_id_no_tag_id_left.c b/tests/lifx/gateway/test_gateway_allocate_tag_id_no_tag_id_left.c
--- a/tests/lifx/gateway/test_gateway_allocate_tag_id_no_tag_id_left.c
+++ b/tests/lifx/gateway/test_gateway_allocate_tag_id_no_tag_id_left.c
@@ -5,6 +5,7 @@
 
 #define MOCKED_LIFX_TAGGING_INCREF
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 static bool tagging_incref_called = false;
 
