# HG changeset patch
# Parent  d69dd060b27b31d808c6ed34c8469f81d326257f

diff --git a/lifx/gateway.c b/lifx/gateway.c
--- a/lifx/gateway.c
+++ b/lifx/gateway.c
@@ -37,7 +37,7 @@
 #include "bulb.h"
 #include "gateway.h"
 #include "broadcast.h"
-#include "timer.h"
+#include "core/timer.h"
 #include "tagging.h"
 #include "core/jsmn.h"
 #include "core/jsonrpc.h"
@@ -57,7 +57,7 @@
     assert(gw);
 
     LGTD_STATS_ADD_AND_UPDATE_PROCTITLE(gateways, -1);
-    lgtd_lifx_timer_stop_timer(gw->refresh_timer);
+    lgtd_timer_stop(gw->refresh_timer);
     event_del(gw->write_ev);
     if (gw->socket != -1) {
         evutil_closesocket(gw->socket);
@@ -239,8 +239,8 @@
 }
 
 static void
-lgtd_lifx_gateway_refresh_callback(struct lgtd_lifx_timer *timer,
-                                   union lgtd_lifx_timer_ctx ctx)
+lgtd_lifx_gateway_refresh_callback(struct lgtd_timer *timer,
+                                   union lgtd_timer_ctx ctx)
 {
     (void)timer;
     struct lgtd_lifx_gateway *gw = ctx.as_ptr;
@@ -252,7 +252,7 @@
 {
     assert(gw);
 
-    lgtd_lifx_timer_activate_timer(gw->refresh_timer);
+    lgtd_timer_activate(gw->refresh_timer);
 }
 
 static struct lgtd_lifx_bulb *
@@ -323,9 +323,9 @@
     gw->next_req_at = received_at;
     gw->last_pkt_at = received_at;
 
-    union lgtd_lifx_timer_ctx ctx = { .as_ptr = gw };
-    gw->refresh_timer = lgtd_lifx_timer_start_timer(
-        LGTD_LIFX_TIMER_ACTIVATE_NOW,
+    union lgtd_timer_ctx ctx = { .as_ptr = gw };
+    gw->refresh_timer = lgtd_timer_start(
+        LGTD_TIMER_ACTIVATE_NOW,
         LGTD_LIFX_GATEWAY_MIN_REFRESH_INTERVAL_MSECS,
         lgtd_lifx_gateway_refresh_callback,
         ctx
@@ -529,10 +529,10 @@
 
     int latency = LGTD_LIFX_GATEWAY_LATENCY(gw);
     if (latency < LGTD_LIFX_GATEWAY_MIN_REFRESH_INTERVAL_MSECS) {
-        if (!lgtd_lifx_timer_ispending_timer(gw->refresh_timer)) {
+        if (!lgtd_timer_ispending(gw->refresh_timer)) {
             int timeout = LGTD_LIFX_GATEWAY_MIN_REFRESH_INTERVAL_MSECS - latency;
             struct timeval tv = LGTD_MSECS_TO_TIMEVAL(timeout);
-            lgtd_lifx_timer_reschedule_timer(gw->refresh_timer, &tv);
+            lgtd_timer_reschedule(gw->refresh_timer, &tv);
             lgtd_debug(
                 "[%s]:%hu latency is %dms, scheduling next GET_LIGHT_STATE in %dms",
                 gw->ip_addr, gw->port, latency, timeout
diff --git a/lifx/gateway.h b/lifx/gateway.h
--- a/lifx/gateway.h
+++ b/lifx/gateway.h
@@ -79,7 +79,7 @@
     struct event                    *write_ev;
     struct evbuffer                 *write_buf;
     bool                            pending_refresh_req;
-    struct lgtd_lifx_timer          *refresh_timer;
+    struct lgtd_timer               *refresh_timer;
 };
 LIST_HEAD(lgtd_lifx_gateway_list, lgtd_lifx_gateway);
 
diff --git a/tests/lifx/gateway/test_gateway_deallocate_tag_id_from_lifx_network.c b/tests/lifx/gateway/test_gateway_deallocate_tag_id_from_lifx_network.c
--- a/tests/lifx/gateway/test_gateway_deallocate_tag_id_from_lifx_network.c
+++ b/tests/lifx/gateway/test_gateway_deallocate_tag_id_from_lifx_network.c
@@ -4,6 +4,7 @@
 
 #define MOCKED_LIFX_TAGGING_DECREF
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 static bool tagging_decref_called = false;
 
diff --git a/tests/lifx/gateway/test_gateway_enqueue_packet.c b/tests/lifx/gateway/test_gateway_enqueue_packet.c
--- a/tests/lifx/gateway/test_gateway_enqueue_packet.c
+++ b/tests/lifx/gateway/test_gateway_enqueue_packet.c
@@ -1,6 +1,7 @@
 #include "gateway.c"
 
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 int
 main(void)
diff --git a/tests/lifx/gateway/test_gateway_enqueue_packet_ring_full.c b/tests/lifx/gateway/test_gateway_enqueue_packet_ring_full.c
--- a/tests/lifx/gateway/test_gateway_enqueue_packet_ring_full.c
+++ b/tests/lifx/gateway/test_gateway_enqueue_packet_ring_full.c
@@ -1,6 +1,7 @@
 #include "gateway.c"
 
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 int
 main(void)
diff --git a/tests/lifx/gateway/test_gateway_enqueue_packet_ring_wraparound.c b/tests/lifx/gateway/test_gateway_enqueue_packet_ring_wraparound.c
--- a/tests/lifx/gateway/test_gateway_enqueue_packet_ring_wraparound.c
+++ b/tests/lifx/gateway/test_gateway_enqueue_packet_ring_wraparound.c
@@ -1,6 +1,7 @@
 #include "gateway.c"
 
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 int
 main(void)
diff --git a/tests/lifx/gateway/test_gateway_handle_tag_labels.c b/tests/lifx/gateway/test_gateway_handle_tag_labels.c
--- a/tests/lifx/gateway/test_gateway_handle_tag_labels.c
+++ b/tests/lifx/gateway/test_gateway_handle_tag_labels.c
@@ -3,6 +3,7 @@
 #include "gateway.c"
 
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 int
 main(void)
diff --git a/tests/lifx/gateway/test_gateway_update_tag_refcounts.c b/tests/lifx/gateway/test_gateway_update_tag_refcounts.c
--- a/tests/lifx/gateway/test_gateway_update_tag_refcounts.c
+++ b/tests/lifx/gateway/test_gateway_update_tag_refcounts.c
@@ -1,6 +1,7 @@
 #include "gateway.c"
 
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 int
 main(void)
diff --git a/tests/lifx/gateway/test_gateway_write_callback.c b/tests/lifx/gateway/test_gateway_write_callback.c
--- a/tests/lifx/gateway/test_gateway_write_callback.c
+++ b/tests/lifx/gateway/test_gateway_write_callback.c
@@ -3,6 +3,7 @@
 #define MOCKED_EVBUFFER_WRITE_ATMOST
 #define MOCKED_EVBUFFER_GET_LENGTH
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 size_t
 evbuffer_get_length(const struct evbuffer *buf)
diff --git a/tests/lifx/gateway/test_gateway_write_callback_clears_ring_full_flag.c b/tests/lifx/gateway/test_gateway_write_callback_clears_ring_full_flag.c
--- a/tests/lifx/gateway/test_gateway_write_callback_clears_ring_full_flag.c
+++ b/tests/lifx/gateway/test_gateway_write_callback_clears_ring_full_flag.c
@@ -3,6 +3,7 @@
 #define MOCKED_EVBUFFER_WRITE_ATMOST
 #define MOCKED_EVBUFFER_GET_LENGTH
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 size_t
 evbuffer_get_length(const struct evbuffer *buf)
diff --git a/tests/lifx/gateway/test_gateway_write_callback_last_packet_on_ring.c b/tests/lifx/gateway/test_gateway_write_callback_last_packet_on_ring.c
--- a/tests/lifx/gateway/test_gateway_write_callback_last_packet_on_ring.c
+++ b/tests/lifx/gateway/test_gateway_write_callback_last_packet_on_ring.c
@@ -3,6 +3,7 @@
 #define MOCKED_EVBUFFER_WRITE_ATMOST
 #define MOCKED_EVBUFFER_GET_LENGTH
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 size_t
 evbuffer_get_length(const struct evbuffer *buf)
diff --git a/tests/lifx/gateway/test_gateway_write_callback_partial_write.c b/tests/lifx/gateway/test_gateway_write_callback_partial_write.c
--- a/tests/lifx/gateway/test_gateway_write_callback_partial_write.c
+++ b/tests/lifx/gateway/test_gateway_write_callback_partial_write.c
@@ -3,6 +3,7 @@
 #define MOCKED_EVBUFFER_WRITE_ATMOST
 #define MOCKED_EVBUFFER_GET_LENGTH
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 size_t
 evbuffer_get_length(const struct evbuffer *buf)
diff --git a/tests/lifx/gateway/test_gateway_write_callback_ring_wraparound.c b/tests/lifx/gateway/test_gateway_write_callback_ring_wraparound.c
--- a/tests/lifx/gateway/test_gateway_write_callback_ring_wraparound.c
+++ b/tests/lifx/gateway/test_gateway_write_callback_ring_wraparound.c
@@ -3,6 +3,7 @@
 #define MOCKED_EVBUFFER_WRITE_ATMOST
 #define MOCKED_EVBUFFER_GET_LENGTH
 #include "test_gateway_utils.h"
+#include "mock_timer.h"
 
 size_t
 evbuffer_get_length(const struct evbuffer *buf)
